text/0000-do-not-recommend.md

```diff
@@ -0,0 +1,246 @@
+   = note: required because of the requirements on the impl of `diesel::expression::NonAggregate` for `diesel::expression::operators::Eq<a_table::columns::id, &str>`
+   = note: required because of the requirements on the impl of `diesel::query_dsl::filter_dsl::FilterDsl<diesel::expression::operators::Eq<a_table::columns::id, &str>>` for `diesel::query_builder::SelectStatement<a_table::table>`
+```
+
+Nowhere in this output is the *actual* missing trait (`AsExpression`) mentioned,
+nor is the type parameter we care about (`sql_types::Integer`), which is *the
+most important piece of information* ever mentioned.
+
+The final motivation for this attribute is actually to *help* Rust give
+transitive impls when it currently isn't. The only time Rust will recommend
+implementing trait `T` in order to get an implementation of trait `U` is if
+there is only one such impl which could potentially apply to your type that
+would result in that behavior.
+
+For example, Diesel has to provide a special impl to insert more than one row at
+a time on SQLite, which doesn't have the keywords needed to safely do this in a
+single query. However, on older versions of Diesel, if there is something
+missing that causes that insert statement to not be valid, Rust will just give
+up because it doesn't know if you wanted the "normal way to insert a thing" impl
+to apply, or the "insert an iterator on SQLite" impl to apply. In the best case
+this would result in "`InsertStatement<{30 page type}>: ExecuteDsl<Sqlite>` is
+not satisfied", which is not helpful, but at least it's not actively misleading.
+In the worst case it would result in "`YourRandomStruct: Iterator` is not
+satisfied. Perhaps you need to implement it?" which is just complete nonsense.
+
+With this annotation, Rust would know that it should *never* recommend the impl
+related to `Iterators`, and will always give diagnostics as if the "normal way
+to insert a thing" impl were the only one that existed.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Since the diagnostics around this RFC aren't ever mentioned in a guide, I'm not
+sure there would be a guide level explanation, but here goes:
+
+Let's imagine you have the following traits:
+
+```
+pub trait Foo {
+}
+
+pub trait Bar {
+}
+
+impl<T: Foo> Bar for T {
+}
+```
+
+If you tried to call a function that expects `T: Bar` with a type that does not
+implement `Bar`, Rust will helpfully notice that if `T` implemented `Foo`, it
+would also implement `Bar`. Because of that, it will recommend that you
+implement `Foo` instead of `Bar`.
+
+This is usually the desired behavior, but in some cases it can result in
+confusing error messages. Perhaps when a function expects `Bar` and it's not
+implemented, it would never make sense to implement `Foo` for that type. In this
+case, we can put `#[do_not_recommend]` above our impl, and Rust will *never*
+recommend implementing `Foo` as a way to get to `Bar`.
```

Comment:
Little bikeshed here: the Guide-level explanation has an illustrating code example of the code we _don't_ want, but not of the code we _want_. That would be:

```
#[do_not_recommend]
pub trait Foo {
}

pub trait Bar {
}

impl<T: Foo> Bar for T {
}
```

Is that right?

Reply:
I think the `#[do_not_recommend]` is supposed on the impl, not on the trait. AFAICT, it is supposed to convey "don't recommend implementing Foo for this impl block to apply so you then get Bar", not "never recommend implementing Foo to get any other impl".

Reply:
Oh, right. And there I am, accidentally showing why it's useful to put the usage example in the guide ;).